pkg/agent/agent.go

```diff
@@ -635,56 +640,70 @@ var allowedMetricsSuffixes = []string{"upstream_rq_total", "upstream_rq_time", "
 
 // MetricsRelayHandler is invoked as a callback when the agent receive metrics from Envoy (sink).
 func (a *Agent) MetricsRelayHandler(
-	logCtx context.Context,
+	ctx context.Context,
 	in *envoyMetrics.StreamMetricsMessage,
 ) {
-	a.metricsRelayMutex.Lock()
-	defer a.metricsRelayMutex.Unlock()
-
 	metrics := in.GetEnvoyMetrics()
-	metricCount := len(metrics)
-
-	if !time.Now().After(a.metricsBackoffUntil) {
-		dlog.Debugf(logCtx, "Drop %d metric(s); next push scheduled for %s",
-			metricCount, a.metricsBackoffUntil.String())
-		return
-	}
 
 	if a.comm != nil && !a.reportingStopped {
 
-		dlog.Infof(logCtx, "Received %d metric(s)", metricCount)
-
 		a.ambassadorAPIKeyMutex.Lock()
 		apikey := a.ambassadorAPIKey
 		a.ambassadorAPIKeyMutex.Unlock()
 
-		outMetrics := make([]*io_prometheus_client.MetricFamily, 0, len(metrics))
+		newMetrics := make([]*io_prometheus_client.MetricFamily, 0, len(metrics))
 
 		for _, metricFamily := range metrics {
 			for _, suffix := range allowedMetricsSuffixes {
 				if strings.HasSuffix(metricFamily.GetName(), suffix) {
-					outMetrics = append(outMetrics, metricFamily)
+					newMetrics = append(newMetrics, metricFamily)
 					break
 				}
 			}
 		}
 
-		outMessage := &agent.StreamMetricsMessage{
-			Identity:     a.agentID,
-			EnvoyMetrics: outMetrics,
+		p, ok := peer.FromContext(ctx)
```

Comment:
Can you move this up above line 650 (`a.ambassadorAPIKeyMutex.Lock()`). 

Since it is required, we can avoid obtaining a lock on the apiKey and not allocate newMetrics if we are just going to return.